validator/json_schema_validation.py

```diff
@@ -0,0 +1,95 @@
+"""Module to run JSON schema validation of a USI submission metadata file against the ArrayExpress submission schema"""
+
+import os
+import jsonschema
+
+import json_schemas
+from utils.converter_utils import read_json_file
+from utils.common_utils import create_logger
+
+
+def validate_submission_json(json_file, schema_file=None, logger=None):
+    """Match a JSON object against a JSON schema and return a human-readable list of all validation errors."""
+
+    if not logger:
+        # Create local logger in the directory of the JSON data file
+        logger = create_logger(os.path.dirname(json_file), "json_validation", os.path.basename(json_file),
+                               logger_name="JSON")
+    if not schema_file:
+        # If no other schema is given, load the schema describing a full ArrayExpress submission
+        schema_file, schema = load_arrayexpress_submission_schema()
+    else:
+        schema = read_json_file(schema_file)
+
+    json_data = read_json_file(json_file)
+
+    # Create validator with 'resolver' to help locate the referenced sub-schemas when interpreting $ref values
+    # in the submission schema (creates absolute paths to the 'submittable' schema files)
+    resolver = jsonschema.RefResolver("file://" + schema_file, schema)
+    validator = jsonschema.Draft4Validator(schema, resolver=resolver)
+
+    # Check validity of schema first
+    # We shouldn't need this but for now, running this test before calling the checks on the object
+    schema_errors = validator.check_schema(schema)
+
+    # Validate the submission JSON against the submission schema
+    validation_errors = validator.iter_errors(json_data)
+
+    # Print out all error messages with where and why details
+    for e in validation_errors:
+        logger.error(format_json_error_message(e))
+
+
+def load_arrayexpress_submission_schema():
+    """Find and load the ArrayExpress submission schema as JSON"""
+
+    # Construct path to schema and load as json
+    schema_path = os.path.dirname(json_schemas.__file__)
+    schema_file = os.path.join(schema_path, "arrayexpress_submission_schema.json")
+    schema = read_json_file(schema_file)
+
+    return schema_file, schema
+
+
```

Comment:
I don't know out of memory, but I would be surprised if there are no good libraries to deal with tracebacks traversals and error formatting for reporting. Only if you find something adequate of course.

Reply:
Ok, will have a look. The traceback here is a list of the elements in the JSON before the element with error(s).